ostd/src/task/scheduler/mod.rs

```diff
@@ -0,0 +1,262 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Scheduling subsystem (in-OSTD part).
+//!
+//! This module defines what OSTD expects from a scheduling implementation
+//! and provides useful functions for controlling the execution flow.
+#![allow(dead_code)]
+
+mod fifo_scheduler;
+
+pub use fifo_scheduler::FifoScheduler;
+use spin::Once;
+
+use super::{processor, task::Task, TaskStatus};
+use crate::{
+    arch::{self, timer},
+    prelude::*,
+};
+
+static SCHEDULER: Once<&'static dyn Scheduler<Task>> = Once::new();
+
+static SHOULD_PREEMPT: ShouldPreemptFlag = ShouldPreemptFlag::new();
+
+/// Injects a scheduler implementation into framework.
+///
+/// This function can only be called once and must be called during the initialization of kernel.
+pub fn inject_scheduler(scheduler: &'static dyn Scheduler<Task>) {
+    SCHEDULER.call_once(|| scheduler);
+
+    timer::register_callback(|| {
+        SCHEDULER.get().unwrap().local_mut_rq_with(&mut |local_rq| {
+            if local_rq.update_current(UpdateFlags::Tick) {
+                SHOULD_PREEMPT.set();
+            }
+        })
+    });
+}
+
+/// Enqueues a given task.
+///
+/// This should only be called with newly-spawned task.
+pub(super) fn enqueue(runnable: Arc<Task>) {
+    // TODO: remove this check for `SCHEDULER`.
+    // Currently OSTD cannot know whether its user has injected a scheduler.
+    if !SCHEDULER.is_completed() {
+        let fifo_scheduler = Box::new(FifoScheduler::default());
+        let static_scheduler = Box::leak(fifo_scheduler);
+        inject_scheduler(static_scheduler);
+    }
+    SCHEDULER
+        .get()
+        .unwrap()
+        .enqueue(runnable, EnqueueFlags::Spawn);
+}
+
+/// Dequeues the current task from its runqueue.
+///
+/// This should only be called if the current is to exit.
+pub(super) fn dequeue() {
+    reschedule(&mut |local_rq| {
+        local_rq.dequeue_current();
+
+        if let Some(next) = local_rq.pick_next_current() {
+            local_rq.set_current(Some(next.clone()));
+            ReschedAction::SwitchTo(next)
+        } else {
+            ReschedAction::Retry
+        }
+    })
+}
+
+/// Tries to make the current task waiting (scheduler part).
+pub(crate) fn try_wait() {
+    reschedule(&mut |local_rq| {
+        local_rq.update_current(UpdateFlags::Wait);
+
+        let current_task = local_rq.dequeue_current().unwrap();
+        let mut current_task_inner = current_task.inner_exclusive_access();
+
+        if current_task_inner.task_status == TaskStatus::Runnable {
```

Comment:
In the latest [commit](bc7c29f444e4396d8f5aeb78e42e91d7d9957116), I refactored the status-related code in `scheduler` module and now all these stuff are done by `LocalRunQueue::pick_next_current` and not bound to any specific use case. I think the current status-related code in `wait` module and `scheduler` module is clear enough, that is, the modification of `task_status` is bound to relevant operations but not a specific module. Of course we can implement a wrapper in `task` module and pass in the same closure every time to achieve your goal but I think it may be unnecessary.